server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java

```diff
@@ -90,25 +95,168 @@ public static TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShar
                         exception
                     );
                     String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
-                    return new TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShard(
-                        allocationId,
-                        shardStateMetadata.primary,
-                        null,
-                        exception
-                    );
+                    return new GatewayShardStarted(allocationId, shardStateMetadata.primary, null, exception);
                 }
             }
 
             logger.debug("{} shard state info found: [{}]", shardId, shardStateMetadata);
             String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
             final IndexShard shard = indicesService.getShardOrNull(shardId);
-            return new TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShard(
+            return new GatewayShardStarted(
                 allocationId,
                 shardStateMetadata.primary,
                 shard != null ? shard.getLatestReplicationCheckpoint() : null
             );
         }
         logger.trace("{} no local shard info found", shardId);
-        return new TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShard(null, false, null);
+        return new GatewayShardStarted(null, false, null);
+    }
+
+    /**
+     * This class encapsulates the metadata about a started shard that needs to be persisted or sent between nodes.
+     * This is used in {@link TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShardsBatch} to construct the response for each node, instead of
+     * {@link TransportNodesListGatewayStartedShards.NodeGatewayStartedShards} because we don't need to save an extra
+     * {@link DiscoveryNode} object like in {@link TransportNodesListGatewayStartedShards.NodeGatewayStartedShards}
+     * which reduces memory footprint of its objects.
+     *
+     * @opensearch.internal
+     */
+    public static class GatewayShardStarted {
```

Comment:
can this be named GatewayStartedShard to keep it consistent?

Reply:
Ack.